api/agent/protocol/cloudevent.go

```diff
@@ -0,0 +1,183 @@
+package protocol
+
+import (
+	"bytes"
+	"context"
+	"encoding/json"
+	"fmt"
+	"io"
+	"net/http"
+	"reflect"
+	"time"
+
+	"go.opencensus.io/trace"
+
+	"github.com/fnproject/fn/api/models"
+)
+
+// CloudEvent is the official JSON representation of a CloudEvent: https://github.com/cloudevents/spec/blob/master/serialization.md
+type CloudEvent struct {
+	CloudEventsVersion string                 `json:"cloudEventsVersion"`
+	EventID            string                 `json:"eventID"`
+	Source             string                 `json:"source"`
+	EventType          string                 `json:"eventType"`
+	EventTypeVersion   string                 `json:"eventTypeVersion"`
+	EventTime          time.Time              `json:"eventTime"` // TODO: ensure rfc3339 format
+	SchemaURL          string                 `json:"schemaURL"`
+	ContentType        string                 `json:"contentType"`
+	Extensions         map[string]interface{} `json:"extensions"`
+	Data               interface{}            `json:"data"` // from docs: the payload is encoded into a media format which is specified by the contentType attribute (e.g. application/json)
+}
+
+type cloudEventIn struct {
+	CloudEvent
+
+	// Deadline string `json:"deadline"`
+	// Protocol CallRequestHTTP `json:"protocol"`
+}
+
+// cloudEventOut the expected response from the function container
+type cloudEventOut struct {
+	CloudEvent
+
+	// Protocol    *CallResponseHTTP `json:"protocol,omitempty"`
+}
+
+// CloudEventProtocol converts stdin/stdout streams from HTTP into JSON format.
+type CloudEventProtocol struct {
+	// These are the container input streams, not the input from the request or the output for the response
+	in  io.Writer
+	out io.Reader
+}
+
+func (p *CloudEventProtocol) IsStreamable() bool {
+	return true
+}
+
+func (h *CloudEventProtocol) writeJSONToContainer(ci CallInfo) error {
+	buf := bufPool.Get().(*bytes.Buffer)
+	buf.Reset()
+	defer bufPool.Put(buf)
+
+	_, err := io.Copy(buf, ci.Input())
+	if err != nil {
+		return err
+	}
+
+	in := cloudEventIn{
+		CloudEvent: CloudEvent{
+			ContentType: ci.ContentType(),
+			EventID:     ci.CallID(),
+			Extensions: map[string]interface{}{
+				// note: protocol stuff should be set on first ingestion of the event in fn2.0, not here
+				"protocol": CallRequestHTTP{
+					Type:       ci.ProtocolType(),
+					Method:     ci.Method(),
+					RequestURL: ci.RequestURL(),
+					Headers:    ci.Headers(),
+				},
+				"deadline": ci.Deadline().String(),
+			},
+		},
+	}
+	if buf.Len() == 0 {
+		// nada
+		// todo: should we leave as null, pass in empty string, omitempty or some default for the content type, eg: {} for json?
+	} else if ci.ContentType() == "application/json" {
+		d := map[string]interface{}{}
```

Comment:
i'd better leave this up to a function instead of doing this on the Fn side.
If you'd take a look at Azure EventGrid event type (which is the closest one to CloudEvents), so that data should be a type of `interface{}`, there would be no need to decode the data until that is necessary by function's needs.

Reply:
The spec says it should be decoded json.

Reply:
so, if client sends a message with content type: application/json, the data has to be a JSON, but should we care? why should we recheck if that is a real json object?

Reply:
We're not rechecking, this just makes sure it's part of the full object that's going into the function and not the escaped string.

Reply:
the spec is pretty clear https://github.com/cloudevents/spec/blob/master/json-format.md#31-special-handling-of-the-data-attribute